### PR TITLE
SRV resolution

### PR DESCRIPTION
**Pull request by jschaul**
* a new shared library for federation-related functions
* implementation of SRV lookup with ordering.

Real integration testing requires creating real DNS records (unless there is a way to mock these?), so I have not included those.

### CODE REVIEW

File: libs/federation-util/src/Network/Federation/Util/Internal.hs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+-- Given a prefix (e.g. _wire) and a domain (e.g. wire.com),
+-- provides a list of A(AAA) names and port numbers upon a successful
+-- DNS-SRV request, or `Nothing' if the DNS-SRV request failed.
+-- Modified version inspired from http://hackage.haskell.org/package/pontarius-xmpp
+srvLookup' :: Text -> Text -> ResolvSeed -> IO (Maybe [(Domain, Word16)])
+srvLookup' prefix realm resolvSeed = withResolver resolvSeed $ \resolver -> do
+    srvResult <- lookupSRV resolver $ encodeUtf8 $ prefix <> "._tcp." <> realm <> "."
+    case srvResult of
+        Right [] -> return Nothing
+        Right [(_, _, _, ".")] -> return Nothing -- explicit "not available"
```
Comment (tiago-loureiro) [inline libs/federation-util/src/Network/Federation/Util/Internal.hs]: > Right [(_, _, _, ".")] -> return Nothing -- explicit "not available"

This is intended to match SRV records such as:

```
0 10 443 .
```

right? Can we add a test for that? Also, what happens if someone sets sth like:
```
0 10 443 .
0 10 443 bla.com.
```

File: libs/federation-util/src/Network/Federation/Util/DNS.hs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+-- >
+-- > main :: IO ()
+-- > main = do
+-- >   rs <- makeResolvSeed defaultResolvConf
+-- >   x <- srvLookup "staging.zinfra.io" rs
+srvLookup :: Text -> ResolvSeed -> IO (Maybe [(Domain, Word16)])
+srvLookup = srvLookup' srvDefaultPrefix
+
+srvDefaultPrefix :: Text
+srvDefaultPrefix = "_wire" -- TODO: _wire-server? _xmpps?
```
Comment (tiago-loureiro) [inline libs/federation-util/src/Network/Federation/Util/DNS.hs]: `_wire-server` sounds 👍

File: libs/federation-util/src/Network/Federation/Util/Internal.hs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        Right srvResult' -> do
+            -- Get [(Domain, PortNumber)] of SRV request, if any.
+            -- Sorts the records based on the priority value.
+            orderedSrvResult <- orderSrvResult srvResult'
+            return $ Just $ fmap (\(_, _, port, domain) -> (domain, port)) orderedSrvResult
+        -- The service is not available at this domain.
+        Left _ -> return Nothing
+
+-- FUTUREWORK: maybe improve sorting algorithm here?
```
Comment (tiago-loureiro) [inline libs/federation-util/src/Network/Federation/Util/Internal.hs]: In terms of performance or logic wise?

File: libs/federation-util/src/Network/Federation/Util/Internal.hs (hunk 1 above)
Comment (jschaul) [reply libs/federation-util/src/Network/Federation/Util/Internal.hs]: This case just deals with the case explicitly mentioned in [the SRV RFC](https://tools.ietf.org/html/rfc2782):

> If there is precisely one SRV RR, and its Target is "." (the root domain), abort.

If someone sets multiple records, one of them being the root domain, well, then just the normal logic applies and a server may not get contacted. We can't prevent people from misconfiguring their DNS.

**Comment by jschaul** [comment]
I added more tests. @tiago-loureiro can you take another look if you're happy with the coverage?

**Pull request merged**